uswds/uswds
# Automate more of the icon collection/build process

Currently, there are a few manual steps to collect our source icons from Material Icons, collect our default icon subset into `usa-icons`, and generate the icon sprite from this collection. It's been easy for these icons to fall behind or for icon source to fall out of sync with the sprite source (`usa-icons`). This PR automates more of the icon collection and creation process, and helps keep everything in sync with less manual effort:

- It uses the npm distribution of Material Icons as its canonical source (`@material-design-icons/svg`)
- It collects the default icon subset as an automated sprite task controlled by a new config file
- It saves all these collected and generated files in `dist` instead of in the `usa-icon` source 

The [Material Icons npm package](https://www.npmjs.com/package/@material-design-icons/svg) is up to date with the most current version of [the Material Icons github repo](https://github.com/google/material-design-icons). Note that per the Google Material Icons repo:

>Google does not currently maintain the npm package for this repo, past v3 (2016). However, user @marella is hosting the following. He tells us these are automatically updated and published using GitHub Actions. Google does NOT monitor or vet these packages.

I've pinned the icons package to its current most-recent version. We can audit the package contents as we consider updating the icon set. But using this package will allow us to better track and stay up to date with any changes to the Material Icons set.

The most recent Material Icons set does not include the following icons from an earlier set that we use in our default set:

- `eco.svg`
- `facebook.svg`
- `info_outline.svg`
- `lightbulb_outline.svg`
- `lock_outline.svg`

Since we still use them, I've added these deprecated icons to a new directory: `usa-icon/src/img/material-icons-deprecated`

- I've removed the Material Icons from `usa-icon` source, since we'll now use the files in the npm package. 
- I've removed the `usa-icons` directory from the `usa-icon` source, since it's now generated as part of the sprite script.
- I've removed the sprite from the `usa-icon` source, since it's a generated file and not source.

All collected/generated files now go straight into `dist` instead of into `usa-icon/src/img`.

## The icon config file
This PR introduces a config file that controls the default icon sprite contents, `packages/usa-icon/src/usa-icons.config.js`:

```js
const config = {
  material: [
    "accessibility_new",
    "accessible_forward",
    ...
  ],
  materialDeprecated: [
    "eco",
    "facebook",
    ...
  ],
  uswds: [
    "bedding",
    "camping",
    ...
  ]
};
```

This config includes the name of each icon in the default set, keyed to its source: Material, deprecated Material, or USWDS. The string is the name only, no filetype. The `sprite` tasks consume this config and output the collected SVGs into `dist/img/usa-icons`.

## How do I...?

**Edit the default icon set.** If you want to edit the contents of the default icon set, edit `usa-icons.config.js`. For instance, if we wanted to add `fax` from the Material Icons set:

```diff
const config = {
  material: [
    "accessibility_new",
    "accessible_forward",
+   "fax"
    ...
  ],
  materialDeprecated: [
    "eco",
    "facebook",
    ...
  ],
  uswds: [
    "bedding",
    "camping",
    ...
  ]
};
```

Run `npx gulp` or `npx gulp buildSprite` to rebuild the sprite.

**Get the newest Material Icons.** If you want to get the newest Material Icons, update the `@material-design-icons/svg` package and run `npx gulp` to rebuild USWDS and create the `dist/img/material-icons` directory.


## How to test
1. Check for proper icon display on the Storybook site.
2. Run `npx gulp buildSpriteStandalone` to test the sprite build script locally
3. Check to see that collected/generated files do not appear in `usa-icon/src/img`, even after building
4. Check that all icon files appear in `dist/img` as expected

Fixes #3889
